Tests/RevenueCatUITests/Templates/ExternalPurchaseAndRestoreTests.swift

```diff
@@ -0,0 +1,393 @@
+            introEligibility: .producing(eligibility: .eligible),
+            purchaseHandler: purchasHandler
+        )
+        .onPurchaseStarted { _ in
+            callbackOrder.append("onPurchaseStarted")
+        }
+        .onPurchaseCompleted { _ in
+            callbackOrder.append("onPurchaseCompleted")
+        }
+        .onPurchaseCancelled {
+            callbackOrder.append("onPurchaseCancelled")
+        }
+        .onPurchaseFailure { _ in
+            callbackOrder.append("onPurchaseFailure")
+        }
+        .onRestoreStarted({
+            callbackOrder.append("onRestoreStarted")
+        })
+        .onRestoreCompleted({ _ in
+            callbackOrder.append("onRestoreCompleted")
+        })
+        .onRestoreFailure({ _ in
+            callbackOrder.append("onRestoreFailure")
+        })
+        .addToHierarchy()
+
+        Task {
+            _ = try await purchasHandler.purchase(package: Self.package)
+            completed = true
+        }
+
+        expect(completed).toEventually(beTrue())
+        expect(callbackOrder).toEventually(equal(["onPurchaseStarted", "performPurchase", "onPurchaseCancelled"]))
+    }
+
+    func testHandleExternalPurchaseFailedPaywall() throws {
+        var callbackOrder = [String]()
+        var passedError: Error?
+
+        let purchasHandler = Self.externalPurchaseHandler { _ in
+            callbackOrder.append("performPurchase")
+            return (userCancelled: false, error: TestError.error)
+        } performRestore: {
+            callbackOrder.append("performRestore")
+            return (success: true, error: nil)
+        }
+
+        try PaywallView(
+            offering: Self.offering.withLocalImages,
+            customerInfo: TestData.customerInfo,
+            introEligibility: .producing(eligibility: .eligible),
+            purchaseHandler: purchasHandler
+        )
+        .onPurchaseStarted { _ in
+            callbackOrder.append("onPurchaseStarted")
+        }
+        .onPurchaseCompleted { _ in
+            callbackOrder.append("onPurchaseCompleted")
+        }
+        .onPurchaseCancelled {
+            callbackOrder.append("onPurchaseCancelled")
+        }
+        .onPurchaseFailure { error in
+            callbackOrder.append("onPurchaseFailure")
+            passedError = error
+        }
+        .onRestoreStarted({
+            callbackOrder.append("onRestoreStarted")
+        })
+        .onRestoreCompleted({ _ in
+            callbackOrder.append("onRestoreCompleted")
+        })
+        .onRestoreFailure({ _ in
+            callbackOrder.append("onRestoreFailure")
+        })
+        .addToHierarchy()
+
+        Task {
+            _ = try await purchasHandler.purchase(package: Self.package)
+        }
+
+        expect(callbackOrder).toEventually(equal(["onPurchaseStarted", "performPurchase", "onPurchaseFailure"]))
+        expect(passedError).toEventually(matchError(TestError.error))
+    }
+
+    func testHandleExternalRestore() throws {
```

Comment:
nitpick: for all of these test names, I'd usually try to include the expectation in the test name, so that when it fails, the broken part jumps out. 

for example for this test, if it fails, does it mean that we just can't handle restores? does it mean that within the handling, we're not calling restore methods in the right order? all of the above? 
If we're testing more than one thing in a single test case it might not be a bad idea to extract the code a bit and split it into multiple separate tests, so that you can for example have 
`testExternalRestoreReturnsCorrectCustomerInfo`, `testExternalRestoreCallsMethodsInCorrectOrder`, etc

Reply:
Good point, done!